src/crypto/aead/aes256gcm.rs

```diff
@@ -0,0 +1,127 @@
+    /// between two randomly generated nonces is nonnegligible and repeated nonce
+    /// use will totally destroy the security of this scheme.
+    /// Use [increment_le](self::Nonce::increment_le)
+    /// or [increment_le_inplace](self::Nonce::increment_le_inplace) to increment a local nonce.
+    /// If you are operating in a multi threaded or distributed environment you must use a shared
+    /// atomic counter protocol instead.
+    ///
+    /// THREAD SAFETY: `gen_initial_nonce` is thread-safe provided that you have
+    /// called [init](crate::init) once before using any other function
+    /// from sodiumoxide.
+    pub fn gen_initial_nonce() -> Nonce {
+        gen_random_nonce()
+    }
+
+    #[cfg(test)]
+    mod test {
+        use super::*;
+
+        #[test]
+        fn test_vector_1() {
+            // Test vector from https://tools.ietf.org/html/rfc7714#section-16.2.2
+            let m = &[
+                0x47, 0x61, 0x6c, 0x6c, 0x69, 0x61, 0x20, 0x65, 0x73, 0x74, 0x20, 0x6f, 0x6d, 0x6e,
+                0x69, 0x73, 0x20, 0x64, 0x69, 0x76, 0x69, 0x73, 0x61, 0x20, 0x69, 0x6e, 0x20, 0x70,
+                0x61, 0x72, 0x74, 0x65, 0x73, 0x20, 0x74, 0x72, 0x65, 0x73,
+            ];
+            let ad = &[
+                0x80, 0x40, 0xf1, 0x7b, 0x80, 0x41, 0xf8, 0xd3, 0x55, 0x01, 0xa0, 0xb2,
+            ];
+            let k = Key([
+                0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d,
+                0x0e, 0x0f, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b,
+                0x1c, 0x1d, 0x1e, 0x1f,
+            ]);
+            let n = Nonce([
+                0x51, 0x75, 0x3c, 0x65, 0x80, 0xc2, 0x72, 0x6f, 0x20, 0x71, 0x84, 0x14,
+            ]);
+            let c_expected = &[
+                0x32, 0xb1, 0xde, 0x78, 0xa8, 0x22, 0xfe, 0x12, 0xef, 0x9f, 0x78, 0xfa, 0x33, 0x2e,
+                0x33, 0xaa, 0xb1, 0x80, 0x12, 0x38, 0x9a, 0x58, 0xe2, 0xf3, 0xb5, 0x0b, 0x2a, 0x02,
+                0x76, 0xff, 0xae, 0x0f, 0x1b, 0xa6, 0x37, 0x99, 0xb8, 0x7b, 0x7a, 0xa3, 0xdb, 0x36,
+                0xdf, 0xff, 0xd6, 0xb0, 0xf9, 0xbb, 0x78, 0x78, 0xd7, 0xa7, 0x6c, 0x13,
+            ];
+            let c = seal(m, Some(ad), &n, &k);
+            assert_eq!(&c[..].len(), &c_expected[..].len());
+            assert_eq!(&c[0..44], &c_expected[0..44]);
+        }
+    }
+}
+
+#[cfg(any(
+    doc,
+    all(
+        any(target_arch = "x86", target_arch = "x86_64"),
+        target_feature = "aes",
+        target_feature = "pclmulqdq"
+    )
+))]
+pub use self::aes::*;
+
+/// `is_available` returns true if the current CPU supports aes256gcm and false otherwise.
+///
+/// WARNING: You must call [init](crate::init) before calling this function; if
+/// you do not the result of calling `is_available` is undefined.
+pub fn is_available() -> bool {
+    unsafe { ffi::crypto_aead_aes256gcm_is_available() == 1 }
```

Comment:
Will libsodium fallback to a software implementation if there is no pclmulqdq&aes?

Reply:
No, there's no safe way to implement AES without hardware support, and thankfully libsodium doesn't try. I exposed the is_available check outside of the feature flag gated module to maintain API parity with libsodium. However, the compile time check alone ensures correct API usage and precludes the possibility of a runtime fault.